java/piranha/src/main/java/com/uber/piranha/XPFlagCleaner.java

```diff
@@ -287,30 +338,85 @@ private API getXPAPI(ExpressionTree et) {
-        if (isLiteralTreeAndMatchesFlagName(arg)
-            || isVarSymbolAndMatchesFlagName(argSym)
-            || isSymbolAndMatchesFlagName(argSym)) {
-          MemberSelectTree mst = (MemberSelectTree) mit.getMethodSelect();
-          String methodName = mst.getIdentifier().toString();
-          if (controlMethods.contains(methodName)) {
-            return API.IS_CONTROL;
-          } else if (treatedMethods.contains(methodName)) {
-            return API.IS_TREATED;
-          } else if (deleteMethods.contains(methodName)) {
-            return API.DELETE_METHOD;
-          } else if (treatmentGroupMethods.contains(methodName)) {
-            return API.IS_TREATMENT_GROUP_CHECK;
+  private API getXPAPI(
+      MethodInvocationTree mit, List<Map<String, Object>> methodPropertiesForMethodName) {
+    for (Map<String, Object> currentMethodProperties : methodPropertiesForMethodName) {
+      Integer argumentIndex = getArgumentIndexFromMap(currentMethodProperties);
+
+      // when argumentIndex is specified, if mit's argument at argIndex doesn't match xpFlagName,
+      // skip to next method property map
+      if (argumentIndex != null) {
+        if (argumentIndex < mit.getArguments().size()) {
+          ExpressionTree argTree = mit.getArguments().get(argumentIndex);
+          Symbol argSym = ASTHelpers.getSymbol(argTree);
+          if (!isArgumentMatchesFlagName(argTree, argSym)) {
+            continue;
           }
+        } else {
+          continue;
+        }
+      }
+      MemberSelectTree mst = ((MemberSelectTree) mit.getMethodSelect());
+      // when returnType is specified, check if mst's return type matches it
+      // if it's not a match, skip to next method property map
+      String returnType = getValueStringFromMap(currentMethodProperties, RETURN_TYPE_STRING);
+      if (returnType != null) {
+        if (!isMethodTreeMatchesMethodProperty(mst, RETURN_TYPE_STRING, returnType)) {
+          continue;
+        }
+      }
+      // when receiverType is specified, check if mst's receiver type matches it
+      // if it's not a match, skip to next method property map
+      String receiverType = getValueStringFromMap(currentMethodProperties, RECEIVER_TYPE_STRING);
+      if (receiverType != null) {
+        if (!isMethodTreeMatchesMethodProperty(mst, RECEIVER_TYPE_STRING, receiverType)) {
+          continue;
         }
       }
+      return flagTypeToAPIMap.getOrDefault(
+          (String) currentMethodProperties.get(FLAG_TYPE_KEY), API.UNKNOWN);
     }
     return API.UNKNOWN;
   }
 
+  private static Map<String, API> initializeFlagTypeToAPIMap() {
+    Map<String, API> map = new HashMap<>();
+    map.put("treated", API.IS_TREATED);
+    map.put("control", API.IS_CONTROL);
+    map.put("empty", API.DELETE_METHOD);
+    map.put("treatmentGroup", API.IS_TREATMENT_GROUP_CHECK);
+    return Collections.unmodifiableMap(map);
+  }
+
+  private boolean isMethodTreeMatchesMethodProperty(
```

Comment:
Can we get rid of this function? It is adding an unnecessary string comparison and there is nothing common between the two branches either.